### PR TITLE
#38/select indoor location 5

### PR DESCRIPTION
**Pull request by thomasgauvin**
This PR is for the implemented functionality of story #38, select indoor/outdoor location.

It makes use of a static class that is used to keep application state. Quite useful.

Still needs UI changed for the list view, and testing of course

### CODE REVIEW

File: app/src/main/java/com/example/concordia_campus_guide/Models/Buildings.java [hunk 1]
```diff
@@ -18,6 +18,14 @@ public Buildings(){
         return Buildings;
     }
 
+    public List<Place> getPlaces(){
+        List<Place> toReturn = new ArrayList<Place>();
```
Comment (ferdousara-parvin) [inline app/src/main/java/com/example/concordia_campus_guide/Models/Buildings.java]: Here too, a more descriptive name would be "buildings".

File: app/src/main/java/com/example/concordia_campus_guide/Models/Rooms.java [hunk 2]
```diff
@@ -15,4 +16,12 @@ public Rooms(List<RoomModel> rooms) {
     public List<RoomModel> getRooms() {
         return Rooms;
     }
+
+    public List<Place> getPlaces(){
+        List<Place> toReturn = new ArrayList<Place>();
```
Comment (ferdousara-parvin) [inline app/src/main/java/com/example/concordia_campus_guide/Models/Rooms.java]: Here too, var name could be changed to "rooms"

File: app/src/main/java/com/example/concordia_campus_guide/Models/Floors.java [hunk 3]
```diff
@@ -15,5 +15,11 @@ public Floors(){
         return Floors;
     }
 
-
+    public List<Place> getPlaces(){
+        List<Place> toReturn = new ArrayList<Place>();
```
Comment (ferdousara-parvin) [inline app/src/main/java/com/example/concordia_campus_guide/Models/Floors.java]: Here too, var name could be changed to "floors"

File: app/src/main/java/com/example/concordia_campus_guide/Global/ApplicationState.java [hunk 4]
```diff
@@ -18,6 +18,7 @@
     private Buildings buildings;
```
Comment (ferdousara-parvin) [inline app/src/main/java/com/example/concordia_campus_guide/Global/ApplicationState.java]: Didn't we say that we won't need the ApplicationState anymore?

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/LocationFragment/LocationFragmentViewModel.java [hunk 5]
```diff
@@ -147,7 +150,7 @@ public void addBuildingMarker(GoogleMap map, LatLng centerPos, String buildingLa
                         .anchor(0.5f,0.5f)
                         .alpha(0.90f)
                         //This line should be included whenever we test the UI for the marker:
-                        .title(buildingLabel)
+                        //.title(buildingLabel)
```
Comment (ferdousara-parvin) [inline app/src/main/java/com/example/concordia_campus_guide/Fragments/LocationFragment/LocationFragmentViewModel.java]: Is this the proper way of doing the Ui test? We might forget to uncomment it.

**Comment by ferdousara-parvin** [comment]
Just tested out what you did, amazing job Thomas :) Left a few comments for you.

File: app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java [hunk 6]
```diff
@@ -71,10 +64,12 @@ public boolean onOptionsItemSelected(MenuItem item){
         int id = item.getItemId();
 
         if(id == R.id.search){
-            //TODO onClick of search_activity button
             Intent openSearch = new Intent(MainActivity.this,
                     SearchActivity.class);
 
+            SelectingToFromState.setMyCurrentLocation(locationFragment.getCurrentLocation());
+            openSearch.putExtra("myCurrentLocation", locationFragment.getCurrentLocation());
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java]: good call, that method was created after this

File: app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    public static Place getTo() {
+        return to;
+    }
+
+    public static void setTo(Place to) {
+        SelectingToFromState.to = to;
+    }
+
+    public static Location getMyCurrentLocation() {
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java]: SelectingToFromState makes it accessible statically, without having reference to the location fragment. For example, we cannot access the location fragment from the search activity nor the routes activity. Therefore, the state is used to pass the data between these.

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java [hunk 8]
```diff
@@ -1,19 +1,93 @@
...
+
+            return true;
+        }
+
+        return super.onKeyDown(keyCode, event);
+    }
+
+    private void exitSelectToFrom(){
+        Intent exitSelectToFrom= new Intent(RoutesActivity.this,
+                MainActivity.class);
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java]: I guess the figma states so, but it doesn't "feel" right when we have it that way

File: app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivityViewModel.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    public List<Floor> getFloors(){
+        return floors;
+    }
+
+    public List<RoomModel> getRooms(){
+        return rooms;
+    }
+
+    public List<Place> getAllPlaces(){
+        List<Place> toReturn = new ArrayList<Place>();
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivityViewModel.java]: sure

File: app/src/main/java/com/example/concordia_campus_guide/Models/Buildings.java (hunk 1 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Buildings.java]: sounds good

File: app/src/main/java/com/example/concordia_campus_guide/Models/Floors.java (hunk 3 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Floors.java]: ok

File: app/src/main/java/com/example/concordia_campus_guide/Models/Rooms.java (hunk 2 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Rooms.java]: done

File: app/src/main/java/com/example/concordia_campus_guide/Global/ApplicationState.java (hunk 4 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Global/ApplicationState.java]: we still need application state to load the json

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/LocationFragment/LocationFragmentViewModel.java (hunk 5 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Fragments/LocationFragment/LocationFragmentViewModel.java]: idk, this is not me, this is part of the building marker story

File: app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java [hunk 10]
```diff
@@ -1,23 +1,95 @@
...
+        //search results filtering according to search text
+        searchText.addTextChangedListener(new TextWatcher(){
+            @Override
+            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2){}
+
+            @Override
+            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2){
+                (SearchActivity.this).adapter.getFilter().filter(charSequence);
+            }
+            @Override
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java]: the OnCreate method I feel that we can refactor it in a better way. For example, if you look in the LocationFragment file, we have a method called initComponent() which takes the responsibility of just initiating the components and then been called in the onCreate(). Furthermore,  you can also put the beforeTextChanged, onTextChanged and the aftertextChanged in a separate method n call it in create. What do you think?

File: app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java [hunk 11]
```diff
@@ -28,4 +100,12 @@ public void onClick(View v) {
...
         });
     }
+
+    private String getPlaceType(Place place){
+        if(place instanceof RoomModel) return RoomModel.class.getSimpleName();
+        if(place instanceof Floor) return Floor.class.getSimpleName();
+        if(place instanceof Building) return Building.class.getSimpleName();
+
+        return null;
+    }
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java]: Do you think we can place this method in the viewModel ?

File: app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivityViewModel.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
+package com.example.concordia_campus_guide.Activities;
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivityViewModel.java]: nicely done 👍

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            convertView.setTag(viewHolder);
+
+        } else {
+            viewHolder = (ViewHolder) convertView.getTag();
+        }
+
+        Place item = getItem(position);
+        if (item!= null) {
+            // My layout has only one TextView
+            // do whatever you want with your string
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: do we need to keep these comments ?

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+                results.count = filters.size();
+                results.values = filters;
+            }
+            else{
+                results.count = originalPlaces.size();
+                results.values = originalPlaces;
+            }
+
+            return results;
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: Do you think  it is possible to split the logic of these method into smaller ones so we can write some units tests for it easier ?

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        @Override
+        protected void publishResults(CharSequence constraint, FilterResults results) {
+            places = (ArrayList<Place>) results.values;
+            notifyDataSetChanged();
+
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: extra empty line

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: extra empty line

File: app/src/main/java/com/example/concordia_campus_guide/Database/Daos/RoomDao.java [hunk 17]
```diff
@@ -15,7 +15,7 @@
 
 //    Fix the query here later
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Database/Daos/RoomDao.java]: Do we need this comment anymore?

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragment.java [hunk 18]
```diff
@@ -73,8 +78,9 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
     public void onActivityCreated(@Nullable Bundle savedInstanceState){
         super.onActivityCreated(savedInstanceState);
         mViewModel = ViewModelProviders.of(this).get(com.example.concordia_campus_guide.Fragments.InfoCardFragment.InfoCardFragmentViewModel.class);
-        mViewModel.setBuilding(ApplicationState.getInstance(getContext()).getBuildings().getBuilding(this.buildingCode));
+        mViewModel.setBuilding(this.buildingCode);
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragment.java]: i like the fact you were able to git rid of the context usage 👍  good job here!

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragmentViewModel.java [hunk 19]
```diff
@@ -1,36 +1,32 @@
...
 
+import androidx.annotation.NonNull;
+import androidx.lifecycle.AndroidViewModel;
+
+import com.example.concordia_campus_guide.Database.AppDatabase;
 import com.example.concordia_campus_guide.Models.Building;
+import com.example.concordia_campus_guide.Models.Place;
 
-public class InfoCardFragmentViewModel extends ViewModel {
+public class InfoCardFragmentViewModel extends AndroidViewModel {
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragmentViewModel.java]: Why did you change the extend AndroidViewModel instead of ViewModel? Any particular reason behind this choice ?

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/LocationFragment/LocationFragmentViewModel.java (hunk 5 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Fragments/LocationFragment/LocationFragmentViewModel.java]: I am aware of this issue. unfortunately, I didn't have the time to investigate about it yet .

File: app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+    public static boolean isSelectTo() {
+        return SelectTo;
+    }
+
+    public static void setSelectToToTrue() {
+        QuickSelect = false;
+        SelectFrom = false;
+        SelectTo = true;
+    }
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java]: what is the difference between this method and  setQuickSelectoTrue() ?? Because the context is the same for both.

File: app/src/main/java/com/example/concordia_campus_guide/Models/Building.java [hunk 21]
```diff
@@ -21,7 +22,7 @@
 import java.util.List;
 
 @Entity (tableName = "buildings",
-        primaryKeys = {"building_code"}
+        indices = {@Index(value = "building_code", unique = true)}
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Models/Building.java]: why did you change the building_code from being the primary key?

File: app/src/main/java/com/example/concordia_campus_guide/Models/Place.java [hunk 22]
```diff
@@ -1,20 +1,21 @@
...
+import com.example.concordia_campus_guide.Database.Converters.DoubleListConverter;
+import com.google.android.gms.maps.model.LatLng;
+
+@Entity(tableName = "places",
+        indices = {@Index("id")}
+        )
+public abstract class Place {
 
-    @Ignore
+    @ColumnInfo(name = "id")
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Models/Place.java]: Ok, so this will be considered as the primary key in the buildings table instead right ?

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java [hunk 23]
```diff
@@ -1,19 +1,93 @@
...
+
+        setBackButtonOnClick();
+    }
+
+    public void onClickTo(View v){
+        SelectingToFromState.setSelectToToTrue();
+        openSearchPage();
+    }
+
+    public void onClickFrom(View v){
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java]: was autogenerated, so I just left as convention

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java [hunk 24]
```diff
@@ -1,19 +1,93 @@
...
+        SelectingToFromState.setSelectFromToTrue();
+        openSearchPage();
+    }
+
+    private void setBackButtonOnClick(){
+        ImageButton backButton = (ImageButton)this.findViewById(R.id.routesPageBackButton);
+        backButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) { exitSelectToFrom(); }
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java]: same thing

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java [hunk 25]
```diff
@@ -1,19 +1,93 @@
...
+        });
+    }
+
+    @Override
+    public boolean onKeyDown(int keyCode, KeyEvent event)  {
+        if (keyCode == KeyEvent.KEYCODE_BACK ) {
+            exitSelectToFrom();
+
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java]: do we really need to remove ```extra empty lines```? I don't like to squish my code because it is harder to read

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java [hunk 26]
```diff
@@ -1,19 +1,93 @@
...
+            return true;
+        }
+
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java]: same as above

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+import androidx.lifecycle.AndroidViewModel;
+
+import com.example.concordia_campus_guide.Database.AppDatabase;
+import com.example.concordia_campus_guide.Models.Place;
+
+public class RoutesActivityViewModel extends AndroidViewModel {
+
+    private AppDatabase appDB;
+    private Place from;
+    private Place to;
```
[the remark this answers is not in the document]
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: sure I agree with that

File: app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java (hunk 10 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java]: yep fixing it

File: app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java (hunk 11 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivity.java]: deleted the method

File: app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivityViewModel.java (hunk 12 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/SearchActivityViewModel.java]: thanks :)

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java (hunk 13 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: lol thanks for catching this

Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: actually, I would leave it because this will be useful for UI @ferdousara-parvin

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java (hunk 14 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: I don't see how we could split this method up, its the fitlering function, it only has one functionality

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragment.java (hunk 18 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragment.java]: :D

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragmentViewModel.java (hunk 19 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragmentViewModel.java]: Need constructor for db, that comes with AndroidViewModel

File: app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java (hunk 20 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java]: quick select is when we select for the first time from the first search, select to is when, from the routes page, we decide to change the starting place

File: app/src/main/java/com/example/concordia_campus_guide/Models/Building.java (hunk 21 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Building.java]: If you check out the rest of the code, you will notice that I removed all pk's from entities extending place. Place now uses the id as pk, therefore all those that extend place use the id as pk. this is because we need to call from db with a same method for all (less coupling, and all the same type means it can be stored in the same variable ie: PlaceId)

File: app/src/main/java/com/example/concordia_campus_guide/Models/Place.java (hunk 22 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Place.java]: yes, room, floor, building all use the id as pk now.

since their previous pk (building_code, floor_code, room_code) are unique, we can still use them for our relations

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java]: ok, no problem with that :)

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivity.java]: It is just a habit that I have. I don't mind either way as long as the code is clearly readable

File: app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragmentViewModel.java (hunk 19 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Fragments/InfoCardFragment/InfoCardFragmentViewModel.java]: ok, I see thank you for clarifying that 👍

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java (hunk 13 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: Can you make it as Todo ??

File: app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java (hunk 14 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Adapters/PlaceToSearchResultAdapter.java]: Rereading the method now, I noticed what you mean. It is  good. I might have misunderstood the functionality of the method when I read it the first time.

File: app/src/main/java/com/example/concordia_campus_guide/Models/Building.java (hunk 21 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Building.java]: good call 👍

File: app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java (hunk 20 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java]: Can't we use one of the other methods which had the same content ? Because that's what took my attention. The fact that the content of the methods were the same .

Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Global/SelectingToFromState.java]: ok, I see now thank you for clarification  👍
